awscli/customizations/history/db.py

```diff
@@ -0,0 +1,234 @@
+            # This is just a performance enhancement so it is optional. Not all
+            # systems will have a sqlite compiled with the WAL enabled.
+            pass
+
+    def cursor(self):
+        return self._connection.cursor()
+
+    def commit(self):
+        self._connection.commit()
+
+    @property
+    def row_factory(self):
+        return self._connection.row_factory
+
+    @row_factory.setter
+    def row_factory(self, row_factory):
+        self._connection.row_factory = row_factory
+
+
+class PayloadSerializer(json.JSONEncoder):
+    def _process_CaseInsensitiveDict(self, obj, type_name):
+        return dict(obj)
+
+    def _process_bytes(self, obj, type_name):
+        encoded = base64.b64encode(obj)
+        return ensure_text_type(encoded)
+
+    def _process_datetime(self, obj, type_name):
+        return obj.isoformat()
+
+    def _unknown(self, obj, type_name):
+        return type_name
+
+    def encode(self, obj):
+        # The default method is not called in PY2 where the JSONEncoder thinks
+        # it can handle a bytes object becausae it can't tell it apart from a
+        # str object. In PY3 this is handled by the _process_bytes method.
+        # For PY2 we have to override the encode method where the encoding of
+        # a bytes-like string will fail, catch the UnicodeDecodeError and
+        # then call _process_bytes and retry the encoding.
+        try:
+            encoded = super(PayloadSerializer, self).encode(obj)
+            return encoded
+        except UnicodeDecodeError:
+            b64encoded = self._process_bytes(obj, 'bytes')
+            encoded = self.encode(b64encoded)
+            return encoded
+
+    def default(self, obj):
+        type_name = type(obj).__name__
+        return getattr(
+            self, '_process_%s' % type_name, self._unknown
+        )(obj, type_name)
+
+
+class DatabaseRecordWriter(object):
+    _WRITE_RECORD = """
+    INSERT INTO records(id, request_id, source, event_type, timestamp, payload)
+    VALUES (?,?,?,?,?,?)"""
+    _REQUEST_LIFECYCLE_EVENTS = set(
+        ['API_CALL', 'HTTP_REQUEST', 'HTTP_RESPONSE', 'PARSED_RESPONSE'])
+    _START_OF_REQUEST_LIFECYCLE_EVENT = 'API_CALL'
+
+    def __init__(self, connection=None):
+        if connection is None:
+            connection = DatabaseConnection()
+        self._connection = connection
+        self._request_ids = {}
+        self._identifier = None
+
+    def write_record(self, record):
+        cursor = self._connection.cursor()
+        db_record = self._create_db_record(record)
+        cursor.execute(self._WRITE_RECORD, db_record)
+        self._connection.commit()
+
+    def _get_request_id(self, event_type):
+        if event_type == self._START_OF_REQUEST_LIFECYCLE_EVENT:
+            self._start_http_lifecycle()
+        if event_type in self._REQUEST_LIFECYCLE_EVENTS:
+            thread_id = threading.current_thread().ident
+            request_id = self._request_ids.get(thread_id, '')
```

Comment:
Just for testing if I insert an HTTP record before inserting an API_CALL record then there will be no thread id. This is the same reason I do not check the thread_id in some of those tests, only in the ones where an API_CALL is inserted first.

I can either:
* Make it a direct access with [] and remove all tests that just insert a solitary http related record without first inserting an API_CALL record.
* Leave it
* Go back to what I was originally doing where if there is no request_id for an http record insertion it generates a new one then uses that one. Leaving API_CALL with the same behavior where it overwrites an existing request_id.

In any case the real thing won't use either behavior that is being used for tests so I don't really like either 1 or 3.

Reply:
I think I am fine keeping it. It will also be more robust if there was an issue in the future where for whatever reason we did not previously put a request id. Like we do not want to lose the whole record.